From 6ecc60ad785a6009e56f8cc5f8e8c94dc3f6e5d6 Mon Sep 17 00:00:00 2001
From: Kaan Alsan <alsan@google.com>
Date: Wed, 28 Jan 2026 13:19:16 -0800
Subject: [PATCH] Correctly position vertical tabs dragged at the bottom of the container

This fixes a bug where dragging a tab near the bottom of the container
(e.g. when attaching to a new window) would incorrectly position the tab
at the start of the container.

This CL introduces a new DragPositionHint enum to achieve this, which
will be expanded upon to support better hit-testing when calculating
drag-point targets.

Bug: 476509652
Change-Id: I23832c8017626f5da001a5270f90cde869e91cc6
Fixed: 477302338
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7526153
Reviewed-by: Eshwar Stalin <estalin@chromium.org>
Commit-Queue: Kaan Alsan <alsan@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1576165}
---

--- a/chrome/browser/ui/views/tabs/vertical/vertical_pinned_tab_container_view.cc
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_pinned_tab_container_view.cc
@@ -168,7 +168,7 @@ void VerticalPinnedTabContainerView::Han
     node = split_tab_view->collection_node();
   }
   if (node) {
-    GetDragHandler().HandleDraggedTabsOverNode(*node);
+    GetDragHandler().HandleDraggedTabsOverNode(*node, std::nullopt);
   }
 }
 
--- a/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc
@@ -152,7 +152,8 @@ void VerticalTabDragHandlerImpl::EndDrag
 }
 
 void VerticalTabDragHandlerImpl::HandleDraggedTabsOverNode(
-    const TabCollectionNode& node) {
+    const TabCollectionNode& node,
+    std::optional<DragPositionHint> position_hint) {
   if (!drag_controller_) {
     // Do nothing if the drag is not attached to our context yet (e.g. on the
     // first iteration of the drag loop).
@@ -169,7 +170,7 @@ void VerticalTabDragHandlerImpl::HandleD
       HandleTabDragOverGroup(node);
       break;
     case TabCollectionNode::Type::UNPINNED:
-      HandleTabDragOverUnpinnedContainer(node);
+      HandleTabDragOverUnpinnedContainer(node, position_hint);
       break;
     default:
       NOTREACHED();
@@ -258,7 +259,17 @@ void VerticalTabDragHandlerImpl::HandleT
 }
 
 void VerticalTabDragHandlerImpl::HandleTabDragOverUnpinnedContainer(
-    const TabCollectionNode& node) {
+    const TabCollectionNode& node,
+    std::optional<DragPositionHint> position_hint) {
+  if (position_hint == DragPositionHint::kBottom) {
+    // Move the selected tabs to the end.
+    tab_strip_model_->MoveSelectedTabsTo(
+        tab_strip_model_->count() -
+            tab_strip_model_->selection_model().selected_tabs().size(),
+        std::nullopt);
+    return;
+  }
+
   const tabs::TabInterface* selected_tab =
       *tab_strip_model_->selection_model().selected_tabs().cbegin();
 
--- a/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.h
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.h
@@ -20,6 +20,11 @@
 class TabCollectionNode;
 class TabStripModel;
 
+enum class DragPositionHint {
+  kTop,    // The drag is at the top of the drag target.
+  kBottom  // The drag is at the bottom of the drag target.
+};
+
 // Interface for views to interact with drag handling.
 class VerticalTabDragHandler {
  public:
@@ -34,8 +39,11 @@ class VerticalTabDragHandler {
   // Ends the drag, if started.
   virtual void EndDrag(EndDragReason reason) = 0;
 
-  // Handles tab strip model updates to reflect a drag over a give tab node.
-  virtual void HandleDraggedTabsOverNode(const TabCollectionNode& node) = 0;
+  // Handles tab strip model updates to reflect a drag over a given node.
+  // Position hint is used to determine where the drag is, relative to the node.
+  virtual void HandleDraggedTabsOverNode(
+      const TabCollectionNode& node,
+      std::optional<DragPositionHint> position_hint) = 0;
 
   // Returns the drag context for this handler.
   virtual TabDragContext* GetDragContext() = 0;
@@ -75,7 +83,9 @@ class VerticalTabDragHandlerImpl : publi
   bool ContinueDrag(views::View& event_source_view,
                     const ui::MouseEvent& event) override;
   void EndDrag(EndDragReason reason) override;
-  void HandleDraggedTabsOverNode(const TabCollectionNode& node) override;
+  void HandleDraggedTabsOverNode(
+      const TabCollectionNode& node,
+      std::optional<DragPositionHint> position_hint) override;
   TabDragContext* GetDragContext() override;
   bool IsViewDragging(const views::View& view) const override;
   bool IsDraggingPinnedTabs() const override;
@@ -124,7 +134,9 @@ class VerticalTabDragHandlerImpl : publi
   void HandleTabDragOverTab(const TabCollectionNode& node);
   void HandleTabDragOverSplit(const TabCollectionNode& node);
   void HandleTabDragOverGroup(const TabCollectionNode& node);
-  void HandleTabDragOverUnpinnedContainer(const TabCollectionNode& node);
+  void HandleTabDragOverUnpinnedContainer(
+      const TabCollectionNode& node,
+      std::optional<DragPositionHint> position_hint);
 
   const raw_ref<TabStripModel> tab_strip_model_;
   const raw_ref<TabCollectionNode> root_node_;
--- a/chrome/browser/ui/views/tabs/vertical/vertical_tab_group_view.cc
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_tab_group_view.cc
@@ -271,7 +271,7 @@ void VerticalTabGroupView::HandleTabDrag
     node = split_tab_view->collection_node();
   }
   CHECK(node);
-  GetDragHandler().HandleDraggedTabsOverNode(*node);
+  GetDragHandler().HandleDraggedTabsOverNode(*node, std::nullopt);
 }
 
 BEGIN_METADATA(VerticalTabGroupView)
--- a/chrome/browser/ui/views/tabs/vertical/vertical_unpinned_tab_container_view.cc
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_unpinned_tab_container_view.cc
@@ -169,6 +169,12 @@ void VerticalUnpinnedTabContainerView::U
 void VerticalUnpinnedTabContainerView::HandleTabDragInContainer(
     const gfx::Point point_in_container) {
   const views::ProposedLayout& target_layout = layout_manager_->target_layout();
+  if (point_in_container.y() >= target_layout.host_size.height()) {
+    GetDragHandler().HandleDraggedTabsOverNode(*collection_node_,
+                                               DragPositionHint::kBottom);
+    return;
+  }
+
   views::View* view_at_point =
       GetViewAtPoint(target_layout, point_in_container);
   const TabCollectionNode* node = collection_node_;
@@ -182,7 +188,7 @@ void VerticalUnpinnedTabContainerView::H
     node = split_tab_view->collection_node();
   }
   CHECK(node);
-  GetDragHandler().HandleDraggedTabsOverNode(*node);
+  GetDragHandler().HandleDraggedTabsOverNode(*node, std::nullopt);
 }
 
 VerticalDraggedTabsContainer&
